src/bin/drag_and_drop_listbox.rs

```diff
@@ -0,0 +1,193 @@
+    // drag_begin will be called when we start dragging. drag_begin is where we call rendering the
+    // widget being dragged next to mouse cursor.
+    event_box.connect_drag_begin(drag_begin);
+    // We don't use this, but just to demonstrate that things can be done when the drag ends.
+    event_box.connect_drag_end(drag_end);
+    // See https://developer.gnome.org/gtkmm-tutorial/stable/sec-dnd-signals.html.en
+    // Called when a source is being dragged and the target calls get_data()
+    event_box.connect_drag_data_get(drag_data_get);
+    // Set the same box as a drag destination.
+    event_box.drag_dest_set(DestDefaults::ALL, &drag_targets, DragAction::MOVE);
+    // Called when get_data() returns
+    event_box.connect_drag_data_received(drag_data_received);
+
+    list_box_row
+}
+
+fn drag_begin<W: IsA<Widget>>(w: &W, ctx: &DragContext) {
+    println!("drag_begin({:?})", w);
+
+    // Get containing ListBoxRow and render it next to the cursor while being dragged
+    match w.get_ancestor(ListBoxRow::static_type()) {
+        None => {
+            println!("Can't get ancestor");
+        }
+        Some(ancestor) => {
+            let alloc = ancestor.get_allocation();
+            let surface: cairo::ImageSurface =
+                cairo::ImageSurface::create(cairo::Format::ARgb32, alloc.width, alloc.height)
+                    .unwrap();
+            let surface_: &cairo::Surface = surface.as_ref();
+
+            {
+                ancestor.get_style_context().add_class("drag-icon");
+                let surface_ctx = cairo::Context::new(surface_);
+                ancestor.draw(&surface_ctx);
+                ancestor.get_style_context().remove_class("drag-icon");
+            }
+
+            let (x, y) = w.translate_coordinates(&ancestor, 0, 0).unwrap();
+            surface.set_device_offset(-x as f64, -y as f64);
+            ctx.drag_set_icon_surface(surface_);
+        }
+    }
+}
+
+fn drag_end<W: IsA<Widget>>(w: &W, _ctx: &DragContext) {
+    println!("drag_end({:?})", w);
+}
+
+fn drag_data_get<W: IsA<Widget>>(w: &W, _ctx: &DragContext, sel: &SelectionData, _: u32, _: u32) {
+    println!("drag_data_get({:?}", w);
+
+    // Get containing ListBoxRow index and put that into SelectionData, to be read in
+    // drag_data_received below
+    let list_box_row = w.get_ancestor(ListBoxRow::static_type()).unwrap();
+    // Downcast it
+    let list_box_row = list_box_row.downcast_ref::<ListBoxRow>().unwrap();
+
+    let row_idx = list_box_row.get_index();
+    sel.set(&Atom::intern(""), 32, &row_idx.to_ne_bytes());
+}
+
+fn drag_data_received<W: IsA<Widget>>(
+    target: &W,
+    _ctx: &DragContext,
+    _x: i32,
+    _y: i32,
+    sel: &SelectionData,
+    _info: u32,
+    _time: u32,
+) {
+    println!("drag_data_received({:?})", target);
+
+    // Index of the widget being dragged, as we stored in the SelectionData in drag_data_get()
+    // above.
+    let row_idx: i32 = {
+        let bytes = sel.get_data();
```

Comment:
You probably want to check the `Atom` and `format` fields here to make sure it's the data you expect

Reply:
If you call `get_format()` and `get_target()` on it first, you can compare that to the things you set in `drag_data_get()`. If the values are different then you got data from a different drag event